### PR TITLE
Symbolic ops macros.

### PR DESCRIPTION
**Pull request by alexknvl**
https://github.com/scalaz/scalaz/pull/1411 served as inspiration.
See https://github.com/scalaz/scalaz/pull/1711/files#diff-726e672735d618c9d9bc069c449a4fbeL6 for an example.

### CODE REVIEW

File: base/shared/src/main/scala/scalaz/typeclass/EqSyntax.scala [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
 package typeclass
 
+import scala.language.experimental.macros
+
 trait EqSyntax {
   implicit final class ToEqOps[A: Eq](a: A) {
-    // TODO: macro syntax
-    def ===(b: A): Boolean = implicitly[Eq[A]].equal(a, b)
+    private[typeclass] type Equal
+    def ===(f: A): Boolean = macro meta.SymOps.fa_1[Equal]
```
Comment (hrhino) [inline base/shared/src/main/scala/scalaz/typeclass/EqSyntax.scala]: why uncap it? `def ===(f: A): Boolean = macro meta.SymOps.fa_1[equal]` seems fine to me.

I wish we had literal types in scala 2; `macro meta.SymOps.fa_1["equal"]` would be better still.

Comment (alexknvl) [reply base/shared/src/main/scala/scalaz/typeclass/EqSyntax.scala]: Fixed.

Comment (hrhino) [reply base/shared/src/main/scala/scalaz/typeclass/EqSyntax.scala]: Thanks!

**Comment by alexknvl** [comment]
All credit due to @edmundnoble

**Comment by jdegoes** [comment]
@alexknvl @edmundnoble This looks awesome! However, it would be best to push the constraint to the method for better compiler errors (&, in theory, performance).

**Comment by alexknvl** [comment]
I think it is out of scope for this PR, but otherwise I agree with the compiler error part. Performance should be the same because of the macro.

**Comment by edmundnoble** [comment]
This is so goddamn good.

**Pull request merged**